src/interface.jl

```diff
@@ -169,8 +205,86 @@ struct Cache{Trule,Ty_cache,Ttangents<:Tuple}
     tangents::Ttangents
 end
 
-_copy!!(dst, src) = copy!(dst, src)
-_copy!!(::Number, src::Number) = src
+const supportedcollections = Union{Tuple,NamedTuple}
+
+"""
+    is_user_defined_struct(T)
+
+Required for checking if datatype `T` is a immutable Composite, Mutable Composite type (returns true) or a built in collection type (returns false).
+Helps in identifying user defined struct for recursing over fields correctly. 
+"""
+function is_user_defined_struct(T)
+    return isconcretetype(T) &&
+           !isprimitivetype(T) &&
+           !(T <: supportedcollections) &&
+           !(T <: Array) &&
+           (@static VERSION >= v"1.11" ? !(T <: Memory) : true)
+end
+
+"""
+    __exclude_unsupported_output(y)
+
+Required for the robust design of [`value_and_pullback`](@ref), [`prepare_pullback_cache`](@ref).  
+Ensures that `y` contains no aliasing, circular references, `Ptr`s or non differentiable datatypes. 
+In the forward pass f(args...) output can only return a "Tree" like datastructure with leaf nodes as primitive types.  
+Refer https://github.com/compintell/Mooncake.jl/issues/517#issuecomment-2715202789 and related issue for details.  
+Internally calls [`__exclude_unsupported_output_internal!`](@ref).
+The design is modelled after `zero_tangent`.
+"""
+function __exclude_unsupported_output(y::T) where {T}
+    if (T <: Union{Dict,Set})
+        throw_datastructure_output_error(y)
+    end
+    __exclude_unsupported_output_internal!(y, Set{UInt}())
+    return nothing
+end
+
+"""
+    __exclude_unsupported_output_internal(y::T, address_set::Set{UInt}) where {T}
+
+For checking if output`y` is a valid Mutable/immutable composite or a primitive type.
+Performs a recursive depth first search over the function output `y` with an `isbitstype()` check base case. The visited memory addresses are stored inside `address_set`.
+If the set already contains a newly visited address, it errors out indicating an Alais or Circular reference.
+Also errors out if `y` is or contains a Pointer.
+It is called internally by [`__exclude_unsupported_output(y)`](@ref).
+"""
+function __exclude_unsupported_output_internal!(y::T, address_set::Set{UInt}) where {T}
+    isbitstype(T) && return nothing
+    if objectid(y) in address_set
+        throw_circular_reference_or_alias_error(y)
+    end
+
+    push!(address_set, objectid(y))
+
+    if is_user_defined_struct(T)
+        # recurse over a composite type.
+        for y_sub in fieldnames(T)
+            # isassigned, isdefined are not defined for Tuples, NamedTuples.
+            if !(T <: supportedcollections) && !isdefined(y, y_sub)
```

Comment:
Also, I think that this handles undefined fields nicely. What's the reason that we need to exclude